components/logging/org.wso2.carbon.logging.appender.http/src/main/java/org/wso2/carbon/logging/appender/http/utils/queue/PersistentQueue.java

```diff
@@ -0,0 +1,242 @@
+                tailerBlock = loadNextBlock();
+                if (tailerBlock != null) {
+                    readData = tailerBlock.peekNextItem();
+                }
+            }
+        }
+        return readData != null? deserializeObject(readData) : null;
+    }
+
+    public boolean isEmpty() {
+
+        return queueMetaDataHandler.getAsJsonArray(QUEUE_BLOCK_LIST_KEY).size() == 1
+                && !tailerBlock.hasUnprocessedItems();
+    }
+
+    private void init() throws PersistentQueueException {
+
+        File queueDirectory = new File(this.queueDirectoryPath);
+        if(!queueDirectory.exists() && !queueDirectory.mkdirs()) {
+            throw new PersistentQueueException(
+                    PersistentQueueException.PersistentQueueErrorTypes.QUEUE_DIRECTORY_CREATION_FAILED,
+                    "Error: Unable to create queue directory");
+        }
+        final String QUEUE_METADATA_FILE_NAME = "queue_metadata.pq";
+        this.queueMetaDataHandler = new MetadataFileHandler(
+                this.queueDirectoryPath + "/" + QUEUE_METADATA_FILE_NAME);
+        initMetaData();
+    }
+
+    private void initMetaData() throws PersistentQueueException {
+
+        // handling the initialization without existing metadata file
+        if(!this.queueMetaDataHandler.isInitialized()) {
+            queueMetaDataHandler.addJsonArray(QUEUE_BLOCK_LIST_KEY, new JsonArray());
+            createNewBlock();
+        }
+        loadAppenderBlock();
+        loadTailerBlock();
+        try {
+            currentDiskUsage = calculateDiskUsage();
+        } catch (IOException e) {
+            throw new PersistentQueueException(
+                    PersistentQueueException.PersistentQueueErrorTypes.QUEUE_DISK_USAGE_CALCULATION_FAILED,
+                    "Error: Unable to calculate disk usage.", e);
+        }
+    }
+
+    // appender should be loaded first to allow equality check when tailer is being loaded
+    private void loadAppenderBlock() throws PersistentQueueException {
+
+        JsonArray queueBlocks = this.queueMetaDataHandler.getAsJsonArray(QUEUE_BLOCK_LIST_KEY);
+        String appenderBlockName = queueBlocks.get(queueBlocks.size() - 1).getAsString();
+        this.appenderBlock = QueueBlock.loadBlock(this.queueDirectoryPath, appenderBlockName);
+    }
+
+    private void loadTailerBlock() throws PersistentQueueException {
+
+        JsonArray queueBlocks = this.queueMetaDataHandler.getAsJsonArray(QUEUE_BLOCK_LIST_KEY);
+        String tailerBlockName = queueBlocks.get(0).getAsString();
+        if(appenderBlock.getFileName().equals(tailerBlockName)) {
+            this.tailerBlock = appenderBlock;
+        }
+        else {
+            this.tailerBlock = QueueBlock.loadBlock(this.queueDirectoryPath, tailerBlockName);
+        }
+    }
+
+    private byte[] serializeObject(T obj) throws PersistentQueueException {
+
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
```

Comment:
use a library for serialization and deserialization

Reply:
Use of org.apache.commons.lang.SerializationUtils has caused a class cast issue due to a class loading issue. Hence the implementation logic of the above function is used here which can be later replaced with the org.apache.commons.lang.SerializationUtils functions.